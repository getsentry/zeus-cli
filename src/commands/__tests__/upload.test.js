'use strict';

/* eslint-env jest */

const fs = require('fs');
const command = require('../upload');
const environment = require('../../environment');
const logger = require('../../logger');
const Zeus = require('@zeus-ci/sdk');

jest.mock('fs');
jest.mock('../../environment');
jest.mock('../../logger');

describe('upload command', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    fs.existsSync.mockImplementation(name => /existing.*\.json/.test(name));
    Zeus.instance.uploadArtifact.mockImplementation(() =>
      Promise.resolve({
        id: '9f32e479-0382-43c3-a18e-d35de81c58dc',
        download_url: '/artifacts/9f32e479/download',
      })
    );
  });

  test('uploads the artifact', () => {
    const argv = {
      file: ['existing.json'],
      build: '12345',
      job: '54321',
      type: 'application/json',
      name: 'renamed.json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(Zeus.instance.uploadArtifact.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uploads multiple artifacts', () => {
    const argv = {
      file: ['existing1.json', 'existing2.json'],
      build: '12345',
      job: '54321',
      type: 'application/json',
      name: 'renamed.json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(Zeus.instance.uploadArtifact.mock.calls).toMatchSnapshot();
    });
  });

  test('logs the artifact download URL', () => {
    const argv = {
      file: ['existing.json'],
      build: '12345',
      job: '54321',
      type: 'application/json',
      name: 'renamed.json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      const downloadUrl = 'https://zeus.ci/artifacts/9f32e479/download';
      expect(logger).toHaveBeenCalledWith(expect.stringMatching(downloadUrl));
    });
  });

  test('uses the build ID from the environment', () => {
    environment.buildId = '12345';
    const argv = {
      file: ['existing.json'],
      job: '54321',
      type: 'application/json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(Zeus.instance.uploadArtifact.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses the job ID from the environment', () => {
    environment.jobId = '54321';
    const argv = {
      file: ['existing.json'],
      build: '12345',
      type: 'application/json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(Zeus.instance.uploadArtifact.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses an empty filename by default', () => {
    const argv = {
      file: ['existing.json'],
      build: '12345',
      job: '54321',
      type: 'application/json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(Zeus.instance.uploadArtifact.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('logs an error for missing files', () => {
    const argv = {
      file: ['invalid.json'],
      build: '12345',
      job: '54321',
      type: 'application/json',
    };

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(logger).toHaveBeenCalledWith(expect.stringMatching(/not exist/));
    });
  });

  test('logs errors from the SDK', () => {
    const argv = {
      file: ['existing.json'],
      build: '12345',
      job: '54321',
      type: 'application/json',
    };

    Zeus.instance.uploadArtifact.mockImplementation(() =>
      Promise.reject(new Error('expected failure'))
    );

    expect.assertions(1);
    return command.handler(argv).then(() => {
      expect(logger).toHaveBeenCalledWith(
        expect.stringMatching(/expected failure/)
      );
    });
  });
});
