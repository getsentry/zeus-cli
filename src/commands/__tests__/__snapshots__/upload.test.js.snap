// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload command uploads multiple artifacts 1`] = `
Array [
  Array [
    Object {
      "build": "12345",
      "file": Object {
        "path": "existing1.json",
      },
      "job": "54321",
      "name": "renamed.json",
      "type": "application/json",
    },
  ],
  Array [
    Object {
      "build": "12345",
      "file": Object {
        "path": "existing2.json",
      },
      "job": "54321",
      "name": "renamed.json",
      "type": "application/json",
    },
  ],
]
`;

exports[`upload command uploads the artifact 1`] = `
Array [
  Object {
    "build": "12345",
    "file": Object {
      "path": "existing.json",
    },
    "job": "54321",
    "name": "renamed.json",
    "type": "application/json",
  },
]
`;

exports[`upload command uses an empty filename by default 1`] = `
Array [
  Object {
    "build": "12345",
    "file": Object {
      "path": "existing.json",
    },
    "job": "54321",
    "name": undefined,
    "type": "application/json",
  },
]
`;

exports[`upload command uses the build ID from the environment 1`] = `
Array [
  Object {
    "build": "12345",
    "file": Object {
      "path": "existing.json",
    },
    "job": "54321",
    "name": undefined,
    "type": "application/json",
  },
]
`;

exports[`upload command uses the build ID from the environment 2`] = `
Array [
  Object {
    "build": "12345",
    "file": Object {
      "path": "existing.json",
    },
    "job": "54321",
    "name": undefined,
    "type": "application/json",
  },
]
`;
